src/opm/parser/eclipse/EclipseState/Grid/EclipseGrid.cpp

```diff
@@ -948,23 +948,37 @@ EclipseGrid::EclipseGrid(const Deck& deck, const int * actnum)
         if (!deck.hasKeyword<ParserKeywords::DRV>())
             throw std::logic_error("The current implementation *must* have radial values specified using the DRV keyword");
 
-        if (!deck.hasKeyword<ParserKeywords::DZV>() || !deck.hasKeyword<ParserKeywords::TOPS>())
-            throw std::logic_error("The current implementation *must* have vertical cell size specified using the DZV and TOPS keywords");
+        if (!(deck.hasKeyword<ParserKeywords::DZ>() || deck.hasKeyword<ParserKeywords::DZV>()) || !deck.hasKeyword<ParserKeywords::TOPS>())
+            throw std::logic_error("The vertical cell size must be specified using the DZ or DZV, and the TOPS keywords");
 
         const std::vector<double>& drv     = deck.getKeyword<ParserKeywords::DRV>().getSIDoubleData();
         const std::vector<double>& dthetav = deck.getKeyword<ParserKeywords::DTHETAV>().getSIDoubleData();
-        const std::vector<double>& dzv     = deck.getKeyword<ParserKeywords::DZV>().getSIDoubleData();
         const std::vector<double>& tops    = deck.getKeyword<ParserKeywords::TOPS>().getSIDoubleData();
-        OpmLog::info(fmt::format("\nCreating cylindrical grid from keywords DRV, DTHETAV, DZV and TOPS"));
+        OpmLog::info(fmt::format("\nCreating spiderweb grid from keywords DRV, DTHETAV, DZV and TOPS"));
 
         if (drv.size() != this->getNX())
             throw std::invalid_argument("DRV keyword should have exactly " + std::to_string( this->getNX() ) + " elements");
 
         if (dthetav.size() != this->getNY())
             throw std::invalid_argument("DTHETAV keyword should have exactly " + std::to_string( this->getNY() ) + " elements");
 
-        if (dzv.size() != this->getNZ())
-            throw std::invalid_argument("DZV keyword should have exactly " + std::to_string( this->getNZ() ) + " elements");
+        size_t volume = this->getNX() * this->getNY() * this->getNZ();
+        std::vector<double> dz(volume);
+        if (deck.hasKeyword<ParserKeywords::DZ>()) {
+            const std::vector<double>& dz_deck = deck.getKeyword<ParserKeywords::DZ>().getSIDoubleData();
+            if (dz_deck.size() != volume)
+                throw std::invalid_argument("DZ keyword should have exactly " + std::to_string( volume ) + " elements");
+            dz = dz_deck;
+        } else {
+            const std::vector<double>& dzv = deck.getKeyword<ParserKeywords::DZV>().getSIDoubleData();
+            if (dzv.size() != this->getNZ())
+                throw std::invalid_argument("DZV keyword should have exactly " + std::to_string( this->getNZ() ) + " elements");
+            auto pz = dz.begin();
```

Comment:
This code was a bit foreign - in particular the `inx`construction; maybe use `std::fill`:
```C++
double zk = 0;
auto area = this->getNX() * this->getNY();
for (std::size_t k= 0; k < this->getNZ() {
     zk += dzv[k]; 
     std::fill(dz.begin() + k*area, dz.begin() + (k+1)*area, zk);
}
```

Reply:
That is what you get if you take a C programmer and let him do C++ :smile: Different way of thinking...